### DISCUSSION TITLE
Remarks on the serializer rework

### DISCUSSION THREAD

**Discussion by Tank86** [dawidchyrzynski/arduino-home-assistant#102] [General]
Hi,

First of all, thanks for the v1 library, that I use on some projects and found very usefull, I did myself a couple of modificaitions in my fork.
I keep an eye on the evolution (v2) because of some devices types that I'm interested in.

Here are a couple of missing features that if possible could be added
- Add more lights features in HAlights (i'm looking especially on RGB colors)
- bring back all the variants of the setValue() from HASensors (I found super usefull also the decimal adjustment feature)

thanks,

**Comment by Tank86** [comment]
In fact while looking at the library evolution (and trying it in my current project), I see a couple of bugs. in the deserialisation of the payload for HALight brightness, Temperature, FanSpeed ...

So I fixed it in this commit https://github.com/Tank86/arduino-home-assistant/commit/b4383bb3198af8126b7c3bedd00dacbeb112b976

**Reply by dawidchyrzynski** [reply]
Hi @Tank86,

Thank you for the information. What's the bug exactly?

**Reply by dawidchyrzynski** [reply]
Ok got it. I will fix this today.

**Reply by Tank86** [reply]
@dawidchyrzynski 

The bug is  inside HAFan::handleSpeedCommand() and all handleXXX() that manages a number as string.
In my setup (homeassistant on rpi) it seems that the payload  does not contains the EOL \0 char (e.g. 64 is length of 2 with '6' '4' chars (and garbadge instead of EOL)

I did not check where the issue exaclty comes from (earlier in the lib or inside mqtt lib or even directly in the homeassistant message.
The way I fixed it is to create a **_strToNumberN_** method with takes the lenght instead of computing it from EOL string. (I kept the original method for compatibility).

In my commit https://github.com/Tank86/arduino-home-assistant/commit/b4383bb3198af8126b7c3bedd00dacbeb112b976 I also added
- more setValues() methods (HASensors) from the v1.3 lib
- the HAlights RGB feature (and prepare the other features)
- simplify code (in calculateNumberSize()) where its possible (a fixed lenght array is way faster than a dynamically allocated array)

This commit is tested ok on my side (but is missing unit tests)

**Reply by dawidchyrzynski** [reply]
I have already pushed the fix for the mentioned bug.

Regarding the HASensor, you should use HASensorNumber instead of HASensor for numeric values. The HASensor is no longer supposed to be used for numeric values. It will be described in the documentation. The RGB feature will be added to the HALight soon.

Unfortunately, I can't merge your changes as it introduces some heavy methods (itoa, utoa, dtostrf, String class). The main goal of the library is to keep the flash overhead as small as possible, even at cost of the performance.

**Reply by Tank86** [reply]
nice fast fix,
Ah yes I missed this new deviceClass, smart way to manage floats and precision

a couple of remarks :
- https://github.com/dawidchyrzynski/arduino-home-assistant/blob/serializer-refactor/src/device-types/HASensorNumber.h#L90 states P2 as default, where is P0
- Any reason to have both setValue and SetValueFloat (and not only setValue like in v1.3) ?

**Reply by dawidchyrzynski** [reply]
Well spotted, I will update the docs. Regarding setValueFloat, it's implemented in this way to keep consistency with getter (setValueFloat, setCurrentValueFloat, getCurrentValueFloat).

Btw, if you want to browse the full documentation you can clone the repository and open the `docs/index.html` file in a web browser.

**Reply by dawidchyrzynski** [reply]
@Tank86, The RGB feature is already available in the HALight. Feel free to give it a go.

**Reply by Tank86** [reply]
@dawidchyrzynski, ah nice fast update.
In the mean time I also did some optimisations in my fork, i'll merge yours

I'll found also a "functional" issue, Hasensor does not send value in onMqttconnected() while HaSensorNumber did.  => because the sensors have "no" values at power-on, the lib will send the default values'0' over mqtt. Which is a bit anoying,
I removed the override onMqttConnected of HaSensorNumber, that fixes the issue. see here => https://github.com/Tank86/arduino-home-assistant/commit/68662c3f68096760c43dea899c08c343989dd701
![image](https://user-images.githubusercontent.com/13146079/194595253-aa4a5232-42c8-4870-96e3-a030e9cd6c50.png)

**Reply by dawidchyrzynski** [reply]
Removing the `onMqttConnected` method introduces another bug that may happen when HA restarts while the device is still running. Anyway, I pushed the fix for this.

**Comment by Tank86** [comment]
Looking at the RGB implementation I see that you need to do a strlen after each call of numberToStr.
_Note: the strlen works only because the array is initialised with 0 before call of numberToStr._

May I suggest that the method should returns the char length, this will avoid this strlen and this will also be very usefull because the method does not set '\0' at end of char array.

**Reply by dawidchyrzynski** [reply]
Done
